[ews-build.webkit.org] Add no-failure-limits label

JonWBedard opened the pull request:
#### 2d4bed9a63007b7dcfe77d538d2de7232dc65ae4
<pre>
[ews-build.webkit.org] Add no-failure-limits label
<a href="https://bugs.webkit.org/show_bug.cgi?id=260394">https://bugs.webkit.org/show_bug.cgi?id=260394</a>
rdar://114094657

Reviewed by Aakash Jain.

Add a label to GitHub which will cause EWS to run all tests,
in spite of may failures. The label will also disable retries.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits.handle_pull_request): Add a new &apos;github_labels&apos; property,
consisting of active GitHub labels.
* Tools/CISupport/ews-build/steps.py:
(GitHub): Add &apos;no-failure-limits&apos; label.
(ValidateChange.validate_github): Extract labels from PR json and update github_labels property.
(RunWebKitTests.setLayoutTestCommand): Remove failure limit and retry if no-failure-limit is set.
(RunWebKitTests.buildCommandKwargs): Set maxTime in remote command.
(RunWebKitTests.evaluateCommand): Do not retry if no-failure-limit is set.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/269973@main">https://commits.webkit.org/269973@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25cfe8dfa999cab339dde1e4f3f2bbac0888345d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25698 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21731 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3258 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24077 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22318 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23779 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1238 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26293 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1035 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21262 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-src.https.sub.html (failure)") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25304 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18695 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/23332 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/949 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1389 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3081 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1256 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3323,6 +3323,11 @@ def doStepIf(self, step):
+        classification = self.getProperty('classification', [])
+        if not isinstance(classification, list):
+            classification = []
+        is_initial_wpt_upload = self.getProperty('action', '') == 'opened' and 'Test-import' in classification
```
Ideally, we'd run all newly added tests regardless of where they came from. We already have the logic to identify those for the stress test.

JonWBedard replied:
WPT imports aren't just adding tests, though. They're modifying existing tests as well. And we want baselines on all configurations, not just the configuration covered by stress tests.

aproskuryakov replied:
> WPT imports aren't just adding tests, though. They're modifying existing tests as well.

This is true of how stress test behaves, too.

> And we want baselines on all configurations, not just the configuration covered by stress tests.

I am not saying that people should rely on stress tests. I am saying that instead of special casing WPT imports, this change should fix the problem for all PRs importing or updating many tests. And that we already have most of the necessary code written, because we needed it for stress tests.

JonWBedard replied:
My understanding is that our problem with WPT imports is about the "importing" piece, not about mass test modification. Basically, during a WPT import, the importer hasn't run the tests they're importing and they know that the baselines either don't exist or are wrong. So we need to disable our failure limit so that we get a result for every test, at least on the initial upload. After that, EWS should function as it normally does.

I can't think of a circumstance, other than WPT import, where we want EWS to disable the failure limit. In other cases, I would expect engineers to have specifically run the tests they're modifying, I don't think disabling the failure limit (and retries) would be desirable.

aproskuryakov replied:
We had many requests for the same behavior outside of WPT imports. Not all of them would be solved by looking at changed tests (e.g. render tree rebases when new information is added to output), but I'm pretty certain that WPT isn't the only case.

JonWBedard replied:
Do we think this behavior is common enough to justify a `no-failure-limits` label? That is, if I recall the alternate proposal. We would basically have a `no-failure-limits` label that `git-webkit pr` applies to initial WPT uploads by default, but also, any PR could add that label to opt in to the behavior.

aproskuryakov replied:
My alternative proposal is to look at changed tests, but `no-failure-limits` seems reasonable too.

JonWBedard replied:
I still don't understand your alternative proposal focusing on changed tests. PRs modify layout tests all the time, and we very rarely want to disable retries and remove the failure limit. Even for WPT uploads, that's only something we want to do on the initial upload of a PR. After the baselines have been gathered, we want EWS to return to its normal behavior.

The only way I can see to generalize the underlying goal is to increase the failure limit based on the number of modified tests. That would let WPT uploads get baselines for all modified tests. It also runs the risk of wasting machine time, though, because we couldn't generally disable retries for all changes which modify layout tests.

aproskuryakov replied:
> The only way I can see to generalize the underlying goal is to increase the failure limit based on the number of modified tests.

Yes, this is what I am talking about.

gsnedders replied:
I will note we don't necessarily have an easy way to know what the total modified tests are: if someone modifies testharness.js practically we need to re-run thousands of tests, even if only one file has changed. If we make the default vary based on number of modified tests, we probably still need as escape-hatch both for WPT imports and other cases.

aj062 reviewed (commented):
Large part of this change is just a refactoring (moving around GitHub class related thing). That makes it hard to locate and review the actual logic change.

That refactoring looks fine. Can you please land that refactoring separately? r=me for that refactoring. Also can use f-string for the lines being modified which are using old-style formatting.

emw-apple commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -113,6 +113,12 @@ class GitHub(object):
     BLOCKED_LABEL = 'merging-blocked'
     SKIP_EWS_LABEL = 'skip-ews'
+    NO_FAILURE_LIMITS = 'no-failure-limits'
+    LABELS = [
```
nit: We could put all the label constants in an [`Enum`](https://docs.python.org/3/library/enum.html?highlight=enum#module-enum) and then we wouldn't need to maintain a separate listing.

JonWBedard replied:
I think in this case, that would make things more confusing, because this is part of the `GitHub` class, not the `GitHubLabels` class.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3776,6 +3790,9 @@ def setLayoutTestCommand(self):
         patch_author = self.getProperty('patch_author')
         if patch_author in ['webkit-wpt-import-bot@igalia.com']:
             self.setCommand(self.command + ['imported/w3c/web-platform-tests'])
+        elif GitHub.NO_FAILURE_LIMITS in self.getProperty('labels', []):
```
Nit: probably better to give this new property more explicit name, e.g.: pr_labels / github_labels.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3776,6 +3790,9 @@ def setLayoutTestCommand(self):
             self.setCommand(self.command + ['imported/w3c/web-platform-tests'])
+        elif GitHub.NO_FAILURE_LIMITS in self.getProperty('labels', []):
+            self.setCommand(self.command + ['--no-retry'])
+            self.maxTime = 60 * 90
```
Please verify if maxTime works fine in buildbot while set in this manner, some of the variables (like timeout) need to be explicitly passed in __init__. And in the only place where we currently use maxTime in this code, we pass it in __init__.

JonWBedard replied:
`maxTime` will not work when set in this manner, based on code inspection. We will need to override `buildCommandKwargs`.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3940,15 +3957,20 @@ def evaluateCommand(self, cmd):
                                                 UploadTestResults(),
                                                 ExtractTestResults()])
             self.finished(WARNINGS)
+
```
Nit: is this extra newline intentional? Do we need it?

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3940,15 +3957,20 @@ def evaluateCommand(self, cmd):
-                ReRunWebKitTests(),
             ])
+            if GitHub.NO_FAILURE_LIMITS not in self.getProperty('labels', []):
+                self.build.addStepsAfterCurrentStep([
```
This isn't going to work in the way we expect it to, because of this buildbot bug: https://github.com/buildbot/buildbot/issues/4874 

Also, the uat instance testing url http://ews-build.webkit-uat.org/#/builders/4/builds/21919 doesn't test this part. Would be good to test this logic of uat instance.

The way we typically tackle this is to maintain a variable like steps_to_add and keep adding steps to that list and finally do a single call to self.build.addStepsAfterCurrentStep with that variable.

aj062 commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -2102,6 +2102,41 @@ def test_success_wpt_import_bot(self):
+
+        self.expectRemoteCommands(
+            ExpectShell(workdir='wkdir',
+                        logfiles={'json': self.jsonFileName},
```
Please try to include maxTime here.

aj062 reviewed (changes requested):
Thanks for separating out the refactoring change and making this PR more concise. It in fact allowed me to easily focus on the core part and find a major concern (with self.build.addStepsAfterCurrentStep).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -113,6 +113,12 @@ class GitHub(object):
     REQUEST_MERGE_QUEUE_LABEL = 'request-merge-queue'
     BLOCKED_LABEL = 'merging-blocked'
     SKIP_EWS_LABEL = 'skip-ews'
+    NO_FAILURE_LIMITS = 'no-failure-limits'
```
Nit: `NO_FAILURE_LIMITS` => `NO_FAILURE_LIMITS_LABEL` 
(for consistency with other similar variables)

aj062 reviewed (approved):
code looks good. 
r+ after it's tested.

JonWBedard committed:
[ews-build.webkit.org] Add no-failure-limits label
https://bugs.webkit.org/show_bug.cgi?id=260394
rdar://114094657

Reviewed by Aakash Jain.

Add a label to GitHub which will cause EWS to run all tests,
in spite of may failures. The label will also disable retries.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits.handle_pull_request): Add a new 'github_labels' property,
consisting of active GitHub labels.
* Tools/CISupport/ews-build/steps.py:
(GitHub): Add 'no-failure-limits' label.
(ValidateChange.validate_github): Extract labels from PR json and update github_labels property.
(RunWebKitTests.setLayoutTestCommand): Remove failure limit and retry if no-failure-limit is set.
(RunWebKitTests.buildCommandKwargs): Set maxTime in remote command.
(RunWebKitTests.evaluateCommand): Do not retry if no-failure-limit is set.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/269973@main
